Implement IdleDuration for RedisTokenBucketRateLimiter

dlxeon opened the pull request:
Fix issue in RedisTokenBucketRateLimiter similar to https://github.com/cristipufu/aspnetcore-redis-rate-limiting/issues/61
The current implementation always returns TimeSpan.Zero for [RateLimiter.IdleDuration](https://learn.microsoft.com/en-us/dotnet/api/system.threading.ratelimiting.ratelimiter.idleduration?view=aspnetcore-7.0#system-threading-ratelimiting-ratelimiter-idleduration). When using a partitioned rate limiter this means the rate limiters created for the partitions [never get cleaned up](https://github.com/dotnet/runtime/blob/c3859ad70b3d5d02b252d07c904160eada32d653/src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/DefaultPartitionedRateLimiter.cs#LL229C17-L229C17) and will accumulate.

Behavior before fix: growing memory up until pod limit, causing pod restarts. 100k unique IP/day, causing 100k+ rate limiters in memory. Growing CPU utilization which is caused by .Net trying to find rate limiters to cleanup.
![image](https://github.com/cristipufu/aspnetcore-redis-rate-limiting/assets/14337206/8654e2c5-31f9-42a1-831f-21ca5b9b1533)

Implementation notes:
- Interlocked operations are required to avoid using lock (...) and protect from imprecise IdleDuration responses in situation when same rate limiter is used to handle multiple simultaneous requests
- Unit tests do not cover case when IdleDuration is returned as TimeSpan.Zero. To cover that we need to introduce some mocks library.

dlxeon commented:
@cristipufu Can you help me with that SonarCloud build issue? I'm not sure if I'm doing something wrong or that is something with configuration.

cristipufu commented:
Hi @dlxeon, thanks for your contribution. I need a few days for research and to think a bit more about this

cristipufu commented on src/RedisRateLimiting/TokenBucket/RedisTokenBucketRateLimiter.cs:
```diff
@@ -14,7 +15,12 @@ public class RedisTokenBucketRateLimiter<TKey> : RateLimiter
+        private int _activeRequestsCount;
+        private long _lastActiveTimestamp = Stopwatch.GetTimestamp();
+
+        public override TimeSpan? IdleDuration => Interlocked.CompareExchange(ref _activeRequestsCount, 0, 0) > 0
```
Quick q: why do we need to track the number of active requests? Can't we just rely on the `_idleSince` timestamp?

dlxeon replied:
This is to ensure that during `AcquireAsyncCoreInternal` execution we will never report we are idle when we don't.

Theoretically if we have infrequent usage of that rate limiter and for some reason Redis connection is slow and request takes 10+ seconds, than we will report 10+ seconds to .Net and then there is a chance .Net will decide that our rate limiter instance is idle and needs to be disposed, even though we are waiting for AcquireAsyncCoreInternal to complete

Probably, that's an overkill and in reality Redis shouldn't be that slow, but I didn't want to make such assumptions for common library.

cristipufu commented on src/RedisRateLimiting/TokenBucket/RedisTokenBucketRateLimiter.cs:
```diff
@@ -14,7 +15,12 @@ public class RedisTokenBucketRateLimiter<TKey> : RateLimiter
+
+        public override TimeSpan? IdleDuration => Interlocked.CompareExchange(ref _activeRequestsCount, 0, 0) > 0
+            ? TimeSpan.Zero
+            : TimeSpan.FromTicks(Stopwatch.GetTimestamp() - _lastActiveTimestamp);
```
Can we please change this to something similar to the runtime's [in-memory implementation](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/FixedWindowRateLimiter.cs#L36C9-L36C157)?

- Rename `_lastActiveTimestamp` to `_idleSince`
- Return `null` instead of `TimeSpan.Zero`
- Explicitly convert `Stopwatch` ticks into `TimeSpan` ticks using `TickFrequency`:
```C#
private static readonly double TickFrequency = (double)TimeSpan.TicksPerSecond / Stopwatch.Frequency;
public override TimeSpan? IdleDuration => _idleSince is null ? null : new TimeSpan((long)((Stopwatch.GetTimestamp() - _idleSince) * TickFrequency));
```
- Implement the same for the other rate limiters (fixed/sliding/concurrency)

cristipufu commented on src/RedisRateLimiting/TokenBucket/RedisTokenBucketRateLimiter.cs:
```diff
@@ -55,18 +61,30 @@ public RedisTokenBucketRateLimiter(TKey partitionKey, RedisTokenBucketRateLimite
 
         protected override RateLimitLease AttemptAcquireCore(int permitCount)
         {
+            _lastActiveTimestamp = Stopwatch.GetTimestamp();
```
This is not needed

cristipufu commented:
Hi @dlxeon I want to merge this, can you please rebase?

dlxeon commented:
Hi @cristipufu I've addressed your suggestions for `RedisTokenBucketRateLimiter` (will do remaining later today/tomorrow), then rebased, but there were multiple conflicts, so final history doesn't look really clean.

cristipufu commented:
Hi @dlxeon, thanks a lot for your contribution. Can you please rebase your PR on top of the main branch? (there are some extra commits that shouldn't be here)

dlxeon commented:
@cristipufu I've rebased and force pushed branch. Please check if everything is fine now.

cristipufu commented:
Thanks @dlxeon!
